### ISSUE TITLE
sls file in base environment has priority over directory/init.sls equivalent file in another environment

### ISSUE THREAD

**Issue by rsuarezsoto** [saltstack/salt#46295]
Sorry if this is a known issue or a known behaviour. I've looked for something alike in the docs but didn't find it.

We have a `file_roots` entry for our states like this:
```
file_roots:
  base:
    - /srv/saltstack/salt
  dev:
    - /srv/saltstack/dev/salt
    - /srv/saltstack/dev/formulas/my-formula
    # States from production
    - /srv/saltstack/salt
```
There is a `conf` directory in `/srv/saltstack/dev/formulas/my-formula`, and an `init.sls` file inside it. There is also a `my-formula` directory in `/srv/saltstack/salt`, and a `conf.sls` file in it.

We have this in the top file:

```
dev:
  'dev_minion':
    - my-formula.conf
base:
  'prod_minion':
    - my-formula.conf
```

I would expect that 'dev_minion' used the file `/srv/saltstack/dev/formulas/my-formula/conf/init.sls`, and 'prod_minion' used `/srv/saltstack/salt/my-formula/conf.sls`. But what happens is that both minions use the latter file. It seems that a sls file has priority over a directory with the same name and an `init.sls` file in it.

Is this expected? Am I doing something wrong?

Thanks in advance.

### Versions Report

Salt Version:
           Salt: 2017.7.4
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.2.2
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
 
System Versions:
           dist: Ubuntu 14.04 trusty
         locale: UTF-8
        machine: x86_64
        release: 3.13.0-85-generic
         system: Linux
        version: Ubuntu 14.04 trusty

**Comment by garethgreenaway** [comment]
@rsuarezsoto thanks for the report.  Everything looks right in your file roots and top.sls file, so far I am unable to duplicate the issue.  Do you have another top.sls file that might be matching the minion? In the dev environment?

**Comment by corywright** [comment]
I'm experiencing this same bug after upgrading to 2017.7.4 from 2016.11.7.

We use `file_roots` overlays to deploy datacenter specific files, such as `/etc/hosts`, based on the Salt environment. Here is a snippet from our master config:

```yaml
file_roots:
  base:
    - /srv/states
    - /srv/files/base
  dal:
    - /srv/states
    - /srv/files/dal
    - /srv/files/base
  slt:
    - /srv/states
    - /srv/files/slt
    - /srv/files/base
```

There is a default `/etc/hosts` that resides in the `base` environment, and then for datacenters that need a different file we use the overlaying feature of `file_roots` and place a copy in the appropriate path:

```
# find /srv/files/ -name hosts 
/srv/files/base/common/etc/hosts
/srv/files/slt/common/etc/hosts
```

The `/srv/states/common/files/etchosts.sls` file:

```sls
/etc/hosts:
  file.managed:
    - source: salt://common/etc/hosts
    - template: jinja
```

Before upgrading to 2017.7.4 running `salt '*' state.apply common.files.etchosts` would use the `/srv/files/slt/common/etc/hosts` file in the `slt` environment, and the `/srv/files/base/common/etc/hosts` file for all others.

After upgrading to 2017.7.4 the `base` environment is always used for files where we use overlays.

In fact, in cases where there is *no* file in the `base` and the file only exists in the datacenter overlay, trying to reference the file in the datacenter specific file_root results in Salt not seeing the file at all:

```
$ find /srv/files/ -name swift-rings
/srv/files/slt/swift-rings
```

`swift-rings` is a directory, with all the OpenStack Swift ring files. Notice that there is not one in `/srv/files/base`, it only exists in the `slt` environment.

The following file (`/srv/states/sync-rings.sls`) has the state to copy this directory down to the minions:

```sls
/etc/swift:
  file.recurse:
    - source: salt://swift-rings/etc/swift
```

When I run `state.apply` for this single state it results in the following:

```
$ salt 'sw-proxy02' state.apply sync-rings
sw-proxy02:
----------
          ID: /etc/swift
    Function: file.recurse
      Result: False
     Comment: Recurse failed: none of the specified sources were found
     Started: 15:28:46.220101
    Duration: 33.664 ms
     Changes:   

Summary for sw-proxy02
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:  33.664 ms
```

If I simply move the directory to the `base` file_root then it works:

```
$ mv /srv/files/slt/swift-rings/ /srv/files/base/

$ salt 'sw-proxy02' state.apply sync-rings
sw-proxy02:

Summary for sw-proxy02
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time: 202.340 ms
```

Move it back and it fails again:

```
$ mv /srv/files/base/swift-rings/ /srv/files/slt/

$ salt 'sw-proxy02' state.apply sync-rings
sw-proxy02:
----------
          ID: /etc/swift
    Function: file.recurse
      Result: False
     Comment: Recurse failed: none of the specified sources were found
     Started: 15:29:58.702509
    Duration: 34.148 ms
     Changes:   

Summary for sw-proxy02
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:  34.148 ms
```

This is being run on the salt-master in the `slt` environment.

```
$ salt --versions-report
Salt Version:
           Salt: 2017.7.4
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.5.0
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Dec  4 2017, 14:50:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-104-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

I'm not able to upgrade to 2018.3 yet, but this seems like a regression that would justify the fix being backported to the 2017.7 series as well.
